src/psyclone/psyir/backend/fortran.py

```diff
@@ -744,63 +744,65 @@ def gen_access_stmt(self, symbol_table):
             f"either 'Symbol.Visibility.PUBLIC' or "
             f"'Symbol.Visibility.PRIVATE'\n")
 
-    def gen_routine_access_stmts(self, symbol_table):
+    def gen_access_stmts(self, symbol_table):
         '''
-        Creates the accessibility statements (R518) for any routine symbols
-        in the supplied symbol table.
+        Creates the accessibility statements (R518) for any routine or
+        imported symbols in the supplied symbol table.
 
         :param symbol_table: the symbol table for which to generate \
                              accessibility statements.
         :type symbol_table: :py:class:`psyclone.psyir.symbols.SymbolTable`
 
-        :returns: the accessibility statements for any routine symbols.
+        :returns: the accessibility statements for any routine or imported \
+                  symbols.
         :rtype: str
 
         :raises InternalError: if a Routine symbol with an unrecognised \
                                visibility is encountered.
         '''
-
         # Find the symbol that represents itself, this one will not need
         # an accessibility statement
         try:
             itself = symbol_table.lookup_with_tag('own_routine_symbol')
         except KeyError:
             itself = None
 
-        public_routines = []
-        private_routines = []
+        public_symbols = []
+        private_symbols = []
         for symbol in symbol_table.symbols:
-            if isinstance(symbol, RoutineSymbol):
+            if (isinstance(symbol, RoutineSymbol) or
+                    isinstance(symbol.interface, (UnresolvedInterface,
+                                                  ImportInterface))):
 
                 # Skip the symbol representing the routine where these
                 # declarations belong
-                if symbol is itself:
+                if isinstance(symbol, RoutineSymbol) and symbol is itself:
                     continue
 
-                # It doesn't matter whether this symbol has a local or global
+                # It doesn't matter whether this symbol has a local or import
                 # interface - its accessibility in *this* context is determined
                 # by the local accessibility statements. e.g. if we are
                 # dealing with the declarations in a given module which itself
                 # uses a public symbol from some other module, the
                 # accessibility of that symbol is determined by the
                 # accessibility statements in the current module.
                 if symbol.visibility == Symbol.Visibility.PUBLIC:
-                    public_routines.append(symbol.name)
+                    public_symbols.append(symbol.name)
                 elif symbol.visibility == Symbol.Visibility.PRIVATE:
-                    private_routines.append(symbol.name)
+                    private_symbols.append(symbol.name)
```

Comment:
Would it be worth/safe skipping the statement it this is already the default?
```
if symbol.visibility == Symbol.Visibility.PUBLIC and not symbol_table.default_visibility in [None, Symbol.Visibility.PUBLIC]:
    return self._nindent + "public\n"
elif symbol.visibility == Symbol.Visibility.PRIVATE and symbol_table.default_visibility == Symbol.Visibility.PRIVATE: 
    private_symbols.append(symbol.name)
```
Just to keep it similar to the input code (although that not a primary requirement)

Reply:
I thought about doing that and then decided not to. However, since you raise it I've now done it as I think it's much nicer. Caused quite a few changes in the tests though.